dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java

```diff
@@ -0,0 +1,209 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.membership;
+
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.status.AlreadyExistsException;
+import alluxio.wire.WorkerInfo;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.etcd.jetcd.KeyValue;
+import org.apache.zookeeper.server.ByteBufferInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * MembershipManager backed by configured etcd cluster.
+ */
+public class EtcdMembershipManager implements MembershipManager {
+  private static final Logger LOG = LoggerFactory.getLogger(EtcdMembershipManager.class);
+  private static final String RING_PATH_FORMAT = "/DHT/%s/AUTHORIZED/";
+  private final AlluxioConfiguration mConf;
+  private AlluxioEtcdClient mAlluxioEtcdClient;
+  private String mClusterName;
+  private String mRingPathPrefix = "";
+
+  /**
+   * @param conf
+   * @return EtcdMembershipManager
+   */
+  public static EtcdMembershipManager create(AlluxioConfiguration conf) {
+    return new EtcdMembershipManager(conf);
+  }
+
+  /**
+   * CTOR for EtcdMembershipManager.
+   * @param conf
+   */
+  public EtcdMembershipManager(AlluxioConfiguration conf) {
+    this(conf, AlluxioEtcdClient.getInstance(conf));
+  }
+
+  /**
+   * CTOR for EtcdMembershipManager with given AlluxioEtcdClient client.
+   * @param conf
+   * @param alluxioEtcdClient
+   */
+  public EtcdMembershipManager(AlluxioConfiguration conf, AlluxioEtcdClient alluxioEtcdClient) {
+    mConf = conf;
+    mClusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    mRingPathPrefix = String.format(RING_PATH_FORMAT, mClusterName);
+    mAlluxioEtcdClient = alluxioEtcdClient;
+  }
+
+  @Override
+  public void join(WorkerInfo wkrAddr) throws IOException {
+    WorkerServiceEntity entity = new WorkerServiceEntity(wkrAddr.getAddress());
+    // 1) register to the ring
+    String pathOnRing = new StringBuffer()
+        .append(mRingPathPrefix)
+        .append(entity.getServiceEntityName()).toString();
+    byte[] ret = mAlluxioEtcdClient.getForPath(pathOnRing);
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

Comment:
who are responsible to close this and the next stream?
shall we use `try-with-resource` to prevent leaking?

Reply:
yes we should, fixed